### PR TITLE
mgr/volumes: the 'mode' should honor idempotent subvolume creation

### PR DESCRIPTION
**Pull request by nmshelke**
## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by nmshelke** [comment]
jenkins test make check

**Comment by nmshelke** [comment]
jenkins test make check

File: src/pybind/mgr/volumes/fs/volume.py [hunk 1]
```diff
@@ -192,6 +198,9 @@ def create_subvolume(self, **kwargs):
                                 'quota': size
                             }
                             subvolume.set_attrs(subvolume.path, attrs)
+                            if kwargs['mode']:
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/volume.py:201]: The `mode` should be updated to default mode (0755) on an idempotent operation if `mode` is not supplied. I know this is not followed for `uid` and `gid` but I think it should be done for uid/gid as well. Need to validate if it breaks existing applications (csi) using them before doing so for uid/gid.

Comment (nmshelke) [reply src/pybind/mgr/volumes/fs/volume.py]: Updated code.
If mode is not specified then it will use default mode (0755)

**Comment by ajarr** [comment]
@Madhu-1 can you please also check that this does not break any assumptions made in CephFS CSI? Thanks a lot!

File: src/pybind/mgr/volumes/fs/volume.py [hunk 2]
```diff
@@ -192,6 +192,8 @@ def create_subvolume(self, **kwargs):
                                 'quota': size
                             }
                             subvolume.set_attrs(subvolume.path, attrs)
+                            mode = octal_str_to_decimal_int(kwargs['mode'])
+                            fs_handle.chmod(subvolume.path, mode)
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/volume.py:196]: Agree!

File: qa/tasks/cephfs/test_volumes.py [hunk 3]
```diff
@@ -952,6 +952,50 @@ def test_subvolume_create_idempotence_resize(self):
         # verify trash dir is clean
         self._wait_for_trash_empty()
 
+    def test_subvolume_create_idempotence_mode(self):
+        # create subvolume
+        subvolume = self._generate_random_subvolume_name()
+        self._fs_cmd("subvolume", "create", self.volname, subvolume)
+
+        # get subvolume metadata (Expecting default mode 16877 means octal mode 40755)
+        subvol_info_before = json.loads(self._get_subvolume_info(self.volname, subvolume))
```
Comment (ajarr) [inline qa/tasks/cephfs/test_volumes.py:961]: Why not https://github.com/ceph/ceph/blob/cb5d80bc4ff0145c19f869612e9a8de7e50c5690/qa/tasks/cephfs/test_volumes.py#L1089 that is easier to follow? You wouldn't need a separate comment this way

File: src/pybind/mgr/volumes/fs/volume.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmshelke) [reply src/pybind/mgr/volumes/fs/volume.py]: Done.

File: qa/tasks/cephfs/test_volumes.py (hunk 3 above)
Comment (nmshelke) [reply qa/tasks/cephfs/test_volumes.py]: Done.

**Comment by Madhu-1** [comment]
> @Madhu-1 can you please also check that this does not break any assumptions made in CephFS CSI? Thanks a lot!

Thanks for tagging @ajarr the mode will remain the same for cephcsi for all subvolumes i.e 777 https://github.com/ceph/ceph-csi/blob/d236968bf9feb31020fc1ed30f9f77af4045aaed/internal/cephfs/core/volume.go#L232-L235 I think we are not affected by this

**Comment by kotreshhr** [comment]
> > @Madhu-1 can you please also check that this does not break any assumptions made in CephFS CSI? Thanks a lot!
> 
> Thanks for tagging @ajarr the mode will remain the same for cephcsi for all subvolumes i.e 777 https://github.com/ceph/ceph-csi/blob/d236968bf9feb31020fc1ed30f9f77af4045aaed/internal/cephfs/core/volume.go#L232-L235 I think we are not affected by this

@Madhu-1 Along the same lines, does CephFS CSI uses idempotent subvolume creations to set uid/gid ? In other words,   does it  change uid/gid of the subvolume after the initial creation ?

**Comment by Madhu-1** [comment]
> > > @Madhu-1 can you please also check that this does not break any assumptions made in CephFS CSI? Thanks a lot!
> > 
> > 
> > Thanks for tagging @ajarr the mode will remain the same for cephcsi for all subvolumes i.e 777 https://github.com/ceph/ceph-csi/blob/d236968bf9feb31020fc1ed30f9f77af4045aaed/internal/cephfs/core/volume.go#L232-L235 I think we are not affected by this
> 
> @Madhu-1 Along the same lines, does CephFS CSI uses idempotent subvolume creations to set uid/gid ? In other words, does it change uid/gid of the subvolume after the initial creation ?

at CephCSI we don't set uid/gid when creating the subvolumes.

**Review by ajarr — APPROVED** [review]
Nice! LGTM

**Comment by nmshelke** [comment]
jenkins test windows

**Review by kotreshhr — APPROVED** [review]
lgtm

**Comment by nmshelke** [comment]
Please find teuthology results for `fs:volumes`
http://pulpito.front.sepia.ceph.com/nshelke-2022-03-11_06:07:16-fs:volumes-wip-nshelke-subvolume-idempotent-creation-distro-default-smithi/

**Review by mchangir — COMMENTED** [review]
LGTM

**Pull request merged**